### ISSUE TITLE
ERR_UNKNOWN_FILE_EXTENSION with node 8.9.0 and ECMAScript module syntax

### ISSUE THREAD

**Issue by rogerta** [istanbuljs/nyc#708]
## Expected Behavior
I'd like to get code coverage when I write code (and unit tests) with ECMAScript module syntax.

## Observed Behavior
Error is thrown:
```
Error [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension: /home/rogerta/.node-spawn-wrap-26280-72eb4eed3ba9/node
    at exports.resolve (internal/loader/ModuleRequest.js:126:13)
    at Loader.resolve (internal/loader/Loader.js:48:40)
    at Loader.getModuleJob (internal/loader/Loader.js:76:40)
    at Loader.import (internal/loader/Loader.js:98:28)
    at module.js:457:29
    at Function.Module._load (module.js:458:7)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
```

## Bonus Points! Code (or Repository) that Reproduces Issue
https://github.com/rogerta/node-tap-esm.git

Repro steps:
```
$ git clone https://github.com/rogerta/node-tap-esm.git
$ cd node-tap-esm
$ npm install
$ npm run cesm
```
```
> tap-es6-import@1.0.0 cesm /home/rogerta/src/tap
> nyc -e .mjs node --experimental-modules foo_test.mjs

(node:26286) ExperimentalWarning: The ESM module loader is experimental.
Error [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension: /home/rogerta/.node-spawn-wrap-26280-72eb4eed3ba9/node
    at exports.resolve (internal/loader/ModuleRequest.js:126:13)
    at Loader.resolve (internal/loader/Loader.js:48:40)
    at Loader.getModuleJob (internal/loader/Loader.js:76:40)
    at Loader.import (internal/loader/Loader.js:98:28)
    at module.js:457:29
    at Function.Module._load (module.js:458:7)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
----------|----------|----------|----------|----------|----------------|
File      |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------|----------|----------|----------|----------|----------------|
All files |  Unknown |  Unknown |  Unknown |  Unknown |                |
----------|----------|----------|----------|----------|----------------|
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! tap-es6-import@1.0.0 cesm: `nyc -e .mjs node --experimental-modules foo_test.mjs`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the tap-es6-import@1.0.0 cesm script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/rogerta/.npm/_logs/2017-11-04T03_49_30_607Z-debug.log
```

## Forensic Information

**Operating System:** ```Linux argon 4.8.0-59-generic #64-Ubuntu SMP Thu Jun 29 19:38:34 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux```
**Environment Information:** https://gist.github.com/rogerta/54ac869a4f784a848b7f620cc8ce8d4e

**Comment by bcoe** [comment]
@rogerta `.mjs` files are not yet supported; There's been a pretty significant change to how module loading works, which breaks some of the underlying behavior of `nyc`. I'm working hard on correcting this, and you can track progress here:

https://github.com/nodejs/node/issues/17337

**Comment by rogerta** [comment]
Cool, thanks Benjamin.

**Comment by jdalton** [comment]
@rogerta You might check out [`@std/esm`](https://github.com/standard-things/esm#tips). It works with the [nyc `--require` option](https://github.com/istanbuljs/nyc#require-additional-modules) without the `--experimental-modules` flag.

**Comment by rogerta** [comment]
Thanks John, I'll check it out.  My workaround for now is to compile my typescript code twice: once with ESM syntax for production and another time with CJS syntax for testing and coverage.  This sounds like a better option until nodejs/node#17337 is fixed.

**Comment by coreyfarrell** [comment]
I just updated `tap` in your repo and tested, I'm not getting an error from `npm run cesm`.  That said I'm also not getting coverage.  For now I recommend using `nyc -i esm` as suggested by @jdalton.  `node --experimental-modules` is just what it says - experimental.  I've opened #1013 as a separate issue to look into why nyc isn't doing anything in that mode, I suspect the new module loader is bypassing the CJS loader system so our hooks are not working, `nyc --show-process-tree=true` does record information about the sub-process so I'm pretty sure that means `spawn-wrap` is still running and we're hooking the CJS system, but the ESM system is bypassing our hooks.
